net/portmapper/portmapper.go

```diff
@@ -241,12 +251,52 @@ var (
 	ErrGatewayNotFound       = errors.New("failed to look up gateway address")
 )
 
-// CreateOrGetMapping either creates a new mapping or returns a cached
+// GetCachedMappingOrStartCreatingOne quickly returns with our current cached portmapping, if any.
+// If there's not one, it starts up a background goroutine to create one.
+// If the background goroutine ends up creating one, the onChange hook registered with the
+// NewClient constructor (if any) will fire.
+func (c *Client) GetCachedMappingOrStartCreatingOne() (external netaddr.IPPort, ok bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Do we have an existing mapping that's valid?
+	now := time.Now()
+	if m := c.pmpMapping; m != nil {
+		if now.Before(m.useUntil) {
+			return m.external, true
+		}
+	}
+
+	// Start creating a portmapping, if one isn't already running.
+	if !c.runningCreate {
+		c.runningCreate = true
+		go c.createMapping()
+	}
+
+	return netaddr.IPPort{}, false
+}
+
+func (c *Client) createMapping() {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	defer func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.runningCreate = false
+	}()
+
+	if _, err := c.createOrGetMapping(ctx); err == nil && c.onChange != nil {
+		go c.onChange()
+	}
```

Comment:
Do we want to handle error cases differently? If there are no portmapping services will this still get run/continually get run? The previous behavior was to keep running so doesn't matter a ton but just wanted to mention it

We probably also want to log the error here if it's not a `NoMappingError`

Reply:
Done.